tfexec/providers_lock.go

```diff
@@ -0,0 +1,76 @@
+package tfexec
+
+import (
+	"context"
+	"os/exec"
+)
+
+type providersLockConfig struct {
+	fsMirror  string
+	netMirror string
+	platforms []string
+	providers []string
+}
+
+var defaultProvidersLockOptions = providersLockConfig{}
+
+type ProvidersLockOption interface {
+	configureProvidersLock(*providersLockConfig)
+}
+
+func (opt *FSMirrorOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.fsMirror = opt.fsMirror
+}
+
+func (opt *NetMirrorOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.netMirror = opt.netMirror
+}
+
+func (opt *PlatformOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.platforms = append(conf.platforms, opt.platform)
+}
+
+func (opt *ProviderOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.providers = append(conf.providers, opt.provider)
+}
+
+// ProvidersLock represents the terraform providers lock
+func (tf *Terraform) ProvidersLock(ctx context.Context, opts ...ProvidersLockOption) error {
+	lockCmd := tf.providersLockCmd(ctx, opts...)
+
+	err := tf.runTerraformCmd(ctx, lockCmd)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
```

Comment:
This command [was only introduced in 0.14](https://github.com/hashicorp/terraform/commit/e270291f194043486bc6c46759308ffa8bdec239), and so we should add some additional checks and return a human-readable error for older versions.

See how we perform similar check for `validate` here: https://github.com/hashicorp/terraform-exec/blob/43e54be55ebc3ce07d381660dcafcb45f567e67c/tfexec/validate.go#L15-L19